p2p/server.py

```diff
@@ -82,136 +85,187 @@ def __init__(self,
-            self.logger.info("Setting up NAT portmap...")
-            # This is experimental and it's OK if it fails, hence the bare except.
             try:
-                await self._add_nat_portmap(lifetime)
-            except Exception as e:
-                if (isinstance(e, upnpclient.soap.SOAPError) and
-                        e.args == (718, 'ConflictInMappingEntry')):
-                    # An entry already exists with the parameters we specified. Maybe the router
-                    # didn't clean it up after it expired or it has been configured by other piece
-                    # of software, either way we should not override it.
-                    # https://tools.ietf.org/id/draft-ietf-pcp-upnp-igd-interworking-07.html#errors
-                    self.logger.info("NAT port mapping already configured, not overriding it")
-                else:
-                    self.logger.exception("Failed to setup NAT portmap")
-
-            try:
-                await wait_with_token(asyncio.sleep(lifetime), token=self.cancel_token)
+                # We start with a sleep because our _run() method will setup the initial portmap.
+                await wait_with_token(
+                    asyncio.sleep(self._nat_portmap_lifetime), token=self.cancel_token)
+                await self.add_nat_portmap()
             except OperationCancelled:
                 break
 
-    async def _add_nat_portmap(self, lifetime: int) -> None:
+    async def add_nat_portmap(self) -> None:
+        self.logger.info("Setting up NAT portmap...")
+        # This is experimental and it's OK if it fails, hence the bare except.
+        try:
+            upnp_dev = await self._discover_upnp_device()
+            if upnp_dev is None:
+                return
+            await self._add_nat_portmap(upnp_dev)
+        except Exception as e:
+            if (isinstance(e, upnpclient.soap.SOAPError) and
+                    e.args == (718, 'ConflictInMappingEntry')):
+                # An entry already exists with the parameters we specified. Maybe the router
+                # didn't clean it up after it expired or it has been configured by other piece
+                # of software, either way we should not override it.
+                # https://tools.ietf.org/id/draft-ietf-pcp-upnp-igd-interworking-07.html#errors
+                self.logger.info("NAT port mapping already configured, not overriding it")
+            else:
+                self.logger.exception("Failed to setup NAT portmap")
+
+    def _find_internal_ip_on_device_network(self, upnp_dev: upnpclient.upnp.Device) -> str:
+        for iface in netifaces.interfaces():
+            for family, addresses in netifaces.ifaddresses(iface).items():
+                # TODO: Support IPv6 addresses as well.
+                if family != netifaces.AF_INET:
+                    continue
+                for item in addresses:
+                    network = item['addr'].rsplit('.', 1)[0]
```

Comment:
This line is really cryptic to me.  Mostly just curious about what format this string is in, what we're pulling out of it, and whether we can use existing parsing tools to do it.